# HW1 - Covariance

In the following formulas, $\mathbf{x}_k$ is an m-dimensional point, or feature vector whose values have been quantized.  $\mathbf{\hat{u}}$ is an m-dimensional vector of the means of $\mathbf{x}_k$ along each dimension. $\oplus$ is the outer product.

Mean of n m-dimensional points:$$\displaystyle \mathbf{\hat{u}} = \frac{1}{n}\sum_{k=1}^{n}\mathbf{x}_k$$ Covariance matrix of n m-dimensional points: $$\displaystyle \hat{\Sigma} = \frac{1}{n}\sum_{k=1}^{n}(\mathbf{x}_k-\hat{u}_k)\oplus(\mathbf{x}_k-\hat{u}_k)^t$$

To begin coding, import numpy.

There are three groups of four points (2-dimensional feature vectors), represented by the variables red, green, and blue, each of which are an array of the points.

In [1]:
import numpy as np

In [2]:
cluster = {'red': np.array([[1, 0], [0, 3], [-1, 0], [0, -3]]),
           'green': np.array([[-9, 7], [-7, 9], [-9, 11], [-11, 9]]),
           'blue': np.array([[11, 7], [14, 8], [11, 9], [8, 8]])}
keys = ('red', 'blue', 'green')

Define a function to compute the mean of all feature vectors along each feature dimension.

In [3]:
def mean(_pts):
    """mean of n points (_pts is a 2-dimensional numpy.ndarray)"""
    _n = _pts.shape[0]  # the number of points
    _m = _pts[0].shape[0]  # the dimension of a point
    _u = np.zeros((_m,), np.float64)  # initialize u to zeros
    for _k in range(_n):
        _u += _pts[_k]
    _u /=_n  # divide by n
    return _u

Define a function to compute the covariance matrix of n feature vectors.

In [4]:
def covariance(_pts):
    """covariance of n points (_pts is a 2-dimensional numpy.ndarray) computed using the outer product algorithm"""
    _n = _pts.shape[0]  # n is the number of m-dimensional points
    _m = _pts[0].shape[0]  # the dimension of the points
    _mu = mean(_pts)  # m-dimensional means
    _cvm = np.zeros((_m, _m), np.float64)  # m by m matrix of zeros
    for _k in range(_n):
        _mm = _pts[_k] - _mu  # subtract the means from the point
        _cvm += np.outer(_mm, _mm)  # accumulate the outer product of mm with itself
    _cvm /= _n  # divide by n
    return _cvm;

Compute and display the covariance matrices of each of the three sets of poits.

In [5]:
for key in keys:
    cvm = covariance(cluster[key])
    print(key)
    print(cvm)

red
[[0.5 0. ]
 [0.  4.5]]
blue
[[4.5 0. ]
 [0.  0.5]]
green
[[2. 0.]
 [0. 2.]]


## Eigenvalues and Eigenvectors

### Red (1)

The eigenvalues of a diagonal matrix are the diagonal elements. $\lambda_1=0.5$, $\lambda_2=4.5$.

Eigenvector $\phi_1$ for $\lambda_1=0.5$, , $\phi_1=\begin{pmatrix}e_1\\e_2\end{pmatrix}$ 
$
\begin{pmatrix}0.5 & 0.0 \\ 0.0 & 4.5\end{pmatrix}\begin{pmatrix}e_1 \\ e_2\end{pmatrix}=0.5\begin{pmatrix}e_1 \\ e_2\end{pmatrix}\text{, }
\begin{pmatrix}0.5e_1+0.0e_2 \\ 0.0e_1+4.5e_2\end{pmatrix}=\begin{pmatrix}0.5e_1 \\ 0.5e_2\end{pmatrix}\text{, }\phi_1=\begin{pmatrix}1\\0\end{pmatrix}
$

Check
$
\begin{pmatrix}0.5 & 0.0 \\ 0.0 & 4.5\end{pmatrix}\begin{pmatrix}1 \\ 0\end{pmatrix}=\begin{pmatrix}0.5\\ 0.0\end{pmatrix}=0.5\begin{pmatrix}1 \\ 0\end{pmatrix}
$

Eigenvector $\phi_2$ for $\lambda_2=4.5$, , $\phi_2=\begin{pmatrix}e_1\\e_2\end{pmatrix}$ 
$
\begin{pmatrix}0.5 & 0.0 \\ 0.0 & 4.5\end{pmatrix}\begin{pmatrix}e_1 \\ e_2\end{pmatrix}=4.5\begin{pmatrix}e_1 \\ e_2\end{pmatrix}\text{, }
\begin{pmatrix}0.5e_1+0.0e_2 \\ 0.0e_1+4.5e_2\end{pmatrix}=\begin{pmatrix}4.5e_1 \\ 4.5e_2\end{pmatrix}\text{, }\phi_2=\begin{pmatrix}0\\1\end{pmatrix}
$

Check
$
\begin{pmatrix}0.5 & 0.0 \\ 0.0 & 4.5\end{pmatrix}\begin{pmatrix}0 \\ 1\end{pmatrix}=\begin{pmatrix}0.0\\ 4.5\end{pmatrix}=4.5\begin{pmatrix}0 \\ 1\end{pmatrix}
$

### Blue (2)

The eigenvalues of a diagonal matrix are the diagonal elements. $\lambda_1=4.5$, $\lambda_2=0.5$.


Eigenvector $\phi_1$ for $\lambda_1=4.5$, , $\phi_1=\begin{pmatrix}e_1\\e_2\end{pmatrix}$ 
$
\begin{pmatrix}4.5 & 0.0 \\ 0.0 & 0.5\end{pmatrix}\begin{pmatrix}e_1 \\ e_2\end{pmatrix}=4.5\begin{pmatrix}e_1 \\ e_2\end{pmatrix}\text{, }
\begin{pmatrix}4.5e_1+0.0e_2 \\ 0.0e_1+0.5e_2\end{pmatrix}=\begin{pmatrix}4.5e_1 \\ 4.5e_2\end{pmatrix}\text{, }\phi_1=\begin{pmatrix}1\\0\end{pmatrix}
$

Check
$
\begin{pmatrix}4.5 & 0.0 \\ 0.0 & 0.5\end{pmatrix}\begin{pmatrix}1 \\ 0\end{pmatrix}=\begin{pmatrix}4.5\\ 0.0\end{pmatrix}=4.5\begin{pmatrix}1 \\ 0\end{pmatrix}
$

Eigenvector $\phi_2$ for $\lambda_2=0.5$, , $\phi_2=\begin{pmatrix}e_1\\e_2\end{pmatrix}$ 
$
\begin{pmatrix}4.5 & 0.0 \\ 0.0 & 0.5\end{pmatrix}\begin{pmatrix}e_1 \\ e_2\end{pmatrix}=0.5\begin{pmatrix}e_1 \\ e_2\end{pmatrix}\text{, }
\begin{pmatrix}4.5e_1+0.0e_2 \\ 0.0e_1+0.5e_2\end{pmatrix}=\begin{pmatrix}0.5e_1 \\ 0.5e_2\end{pmatrix}\text{, }\phi_2=\begin{pmatrix}0\\1\end{pmatrix}
$

Check
$
\begin{pmatrix}4.5 & 0.0 \\ 0.0 & 0.5\end{pmatrix}\begin{pmatrix}0 \\ 1\end{pmatrix}=\begin{pmatrix}0.0\\ 0.5\end{pmatrix}=0.5\begin{pmatrix}0 \\ 1\end{pmatrix}
$

### Green (3)

The eigenvalues of a diagonal matrix are the diagonal elements, and in this case have a multiplicity of 2. $\lambda_1=2.0$, $\lambda_2=2.0$.

Eigenvector $\phi_1$ for $\lambda_1=2.0$, , $\phi_1=\begin{pmatrix}e_1\\e_2\end{pmatrix}$ 
$
\begin{pmatrix}2.0 & 0.0 \\ 0.0 & 2.0\end{pmatrix}\begin{pmatrix}e_1 \\ e_2\end{pmatrix}=2.0\begin{pmatrix}e_1 \\ e_2\end{pmatrix}\text{, }
\begin{pmatrix}2.0e_1+0.0e_2 \\ 0.0e_1+2.0e_2\end{pmatrix}=\begin{pmatrix}2.0e_1 \\ 2.0e_2\end{pmatrix}\text{, }\phi_1=\begin{pmatrix}1\\0\end{pmatrix}\text{, }\phi_2=\begin{pmatrix}0 \\ 1\end{pmatrix} 
$

Check
$
\begin{pmatrix}2.0 & 0.0 \\ 0.0 & 2.0\end{pmatrix}\begin{pmatrix}1 \\ 0\end{pmatrix}=\begin{pmatrix}2.0\\ 0.0\end{pmatrix}=2.0\begin{pmatrix}1 \\ 0\end{pmatrix}
\begin{pmatrix}2.0 & 0.0 \\ 0.0 & 2.0\end{pmatrix}\begin{pmatrix}0 \\ 1\end{pmatrix}=\begin{pmatrix}0.0\\ 2.0\end{pmatrix}=2.0\begin{pmatrix}0 \\ 1\end{pmatrix}
$

Note that the equations also yield the eigenvector solution $\phi_3=\begin{pmatrix}1 \\ 1\end{pmatrix}$, which is a linear combination of $\phi_1$ and $\phi_2$, $\phi_3=\phi_1+\phi_2$.